📖 Quickstart / add other sections (#1379)
kubernetes-sigs/cluster-api: docs/book/src/user/quick-start.md

# Quick Start

In this tutorial we'll cover the basics of how to use Cluster API to create one or more Kubernetes clusters.

## Prerequisites

- Install and setup [kubectl] in your local environment.

> - A Kubernetes cluster to be used as [management cluster](../reference/glossary.md#management-cluster), for the purpose of this quick start, we suggest you to use [kind].

Author: @timothysc Suggested to leave kind here because it's a quick start, not a production tutorial

> - A Kubernetes cluster to be used as [management cluster](../reference/glossary.md#management-cluster), for the purpose of this quick start, we suggest you to use [kind].

Reviewer: That is fine, but if we don't explicitly call out that it is for non-production use, someone will use it for more than a local test and complain when it breaks.

> - A Kubernetes cluster to be used as [management cluster](../reference/glossary.md#management-cluster), for the purpose of this quick start, we suggest you to use [kind].

Author: ptal, added ` as a non production-ready solution.`

> - A Kubernetes cluster to be used as [management cluster](../reference/glossary.md#management-cluster), for the purpose of this quick start, we suggest you to use [kind].

Reviewer: Personally, I think we should call it out as a warning, similar to the vsphere provider examples.

> - A Kubernetes cluster to be used as [management cluster](../reference/glossary.md#management-cluster), for the purpose of this quick start, we suggest you to use [kind].

Author: I highlighted a little with some _italic_, folks should know from kind docs that it's not a long term solution imo

    ```

## Installation

Using [kubectl], let's create the components on the management cluster:

#### Install Cluster API Components
```bash
kubectl create -f https://github.com/kubernetes-sigs/cluster-api/releases/latest/download/cluster-api-components.yaml
```

#### Install the Bootstrap Provider Components

{{#tabs name:"tab-installation-bootstrap" tabs:"Kubeadm"}}
{{#tab Kubeadm}}

```bash
kubectl create -f https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/releases/latest/download/bootstrap-components.yaml
```

{{#/tab }}
{{#/tabs }}

#### Install Infrastructure Provider Components

{{#tabs name:"tab-installation-infrastructure" tabs:"AWS"}}
{{#tab AWS}}

For more information about prerequisites, credentials management, and or permissions for AWS, visit the [getting started guide](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/getting-started.md).

```bash
kubectl create -f https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/releases/latest/download/infrastructure-components.yaml
```

{{#/tab }}
{{#/tabs }}

## Usage

The Cluster API resources are now installed.

> ![](../images/management-cluster.svg)

Author: Done, ptal 😊

> ![](../images/management-cluster.svg)

Reviewer: The image still looks the same, but I also realized my comment didn't come through correctly, I was referencing `<<Manager>>`

> ![](../images/management-cluster.svg)

Author: Check now, I changed the svg but didn't push 😂

Now that we've got Cluster API, Bootstrap and Infrastructure resources installed,
let's proceed to create a single node cluster.

For the purpose of this tutorial, we'll name our cluster `capi-quickstart`.

{{#tabs name:"tab-usage-cluster-resource" tabs:"AWS"}}
{{#tab AWS}}

```yaml
---
apiVersion: cluster.x-k8s.io/v1alpha2
kind: Cluster
metadata:
  name: capi-quickstart
spec:
  clusterNetwork:
    pods:

>       cidrBlocks: ["192.168.0.0/16"]

Reviewer: Might be good to add a comment as to why this is here

  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    kind: AWSCluster
    name: capi-quickstart
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: AWSCluster
metadata:
  name: capi-quickstart
spec:

>   region:

Reviewer: Did you intend to leave region empty here?

  sshKeyName: default
```
{{#/tab }}
{{#/tabs }}

> Now that we've created the cluster object, we can create the first control plane Machine.

Reviewer: If we are only going to deploy a single control plane instance for the quick start (which we likely should):
```suggestion
Now that we've created the cluster object, we can create a control plane Machine.
```

{{#tabs name:"tab-usage-controlplane-resource" tabs:"AWS"}}
{{#tab AWS}}

```yaml
---
apiVersion: cluster.x-k8s.io/v1alpha2
kind: Machine
metadata:
  name: capi-quickstart-controlplane-0
  labels:
    cluster.x-k8s.io/control-plane: "true"
    cluster.x-k8s.io/cluster-name: "capi-quickstart"
spec:
  version: v1.15.3
  bootstrap:
    configRef:
      apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
      kind: KubeadmConfig
      name: capi-quickstart-controlplane-0
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    kind: AWSMachine
    name: capi-quickstart-controlplane-0
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: AWSMachine
metadata:
  name: capi-quickstart-controlplane-0
spec:

>   instanceType: t2.

Reviewer: valid instance type here? If going with a burstable instance for the quick start, should likely be a `t3.large`

>   iamInstanceProfile: "controllers.cluster-api-provider-aws.sigs.k8s.io"

Reviewer: This will error if they haven't run `clusterawsadm` to create this profile, we should probably link out to a prerequisites doc for the aws provider prior to this.

Reviewer: Maybe just a comment that this should already exist, and a link to capa docs/prerequisites?

Author: There is a link to the pre-reqs in the installation part

Reviewer: So it is, might want to highlight it a bit more so that it is obvious that if those steps aren't run that the following steps will fail.

>   sshKeyName: default

Reviewer: Maybe a comment here saying that this key needs to already exist in the region they are deploying to?

Author: Added

---
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
kind: KubeadmConfig
metadata:
  name: capi-quickstart-controlplane-0
spec:
  initConfiguration:
    nodeRegistration:

>       name: '{{ ds.meta_data.hostname }}'

Reviewer: Maybe a comment as to why this is necessary.

Author: These details and all the other down should really be in an advanced section of the kubeadm provider, or in another section of this tutorial. Not sure if they're really helpful for this quickstart.

Reviewer: If we don't provide some context then users will be confused as well.

Reviewer: Especially with this one since it's an embedded jinja template in an otherwise normal yaml file.

Author: I tried to write up something about it, but I didn't come up with anything useful, do you have any suggestion? I'm really torn on this because I don't want to make users worry about something that they don't have to mind at this point of their journey in CAPI.

@timothysc any thoughts?

Author: Added a note on all the kubeadm specific configuration to look at the kubeadm docs

Reviewer: This isn't really kubeadm specific, though...  This is very AWS specific in that we need the Node Name to match with a specific value that is used for looking up the instance within the cloud-provider apis when the cloud-provider integration is enabled.

      kubeletExtraArgs:

>         cloud-provider: aws

Reviewer: comment here as well.

Author: These should be clear I think?

Reviewer: Maybe? But if we expect users to generally copy/paste comments only help improve readability.

  clusterConfiguration:
    apiServer:
      extraArgs:

>         cloud-provider: aws

Reviewer: comment here as well.

    controllerManager:
      extraArgs:

>         cloud-provider: aws

Reviewer: comment here as well.

```
{{#/tab }}
{{#/tabs }}

> 

Reviewer: Maybe also add a MachineDeployment creation to this as well?

Author: Maybe in a later tutorial? I think the QS should only include a single machine for now, we can do a sub-section to create a MachineDeployment in addition

Reviewer: If we defer it, then users wouldn't be able to deploy any type of test application, since the control plane nodes are tainted by default.

Author: Added, ptal

```bash
kubectl create -f https://github.com/kubernetes-sigs/cluster-api/releases/download/v0.2.1/cluster-api-components.yaml
```

#### Install the Bootstrap Provider Components

{{#tabs name:"tab-installation-bootstrap" tabs:"Kubeadm"}}
{{#tab Kubeadm}}

Check the [Kubeadm provider releases](https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/releases) for an up-to-date components file.

```bash
kubectl create -f https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/releases/download/v0.1.0/bootstrap-components.yaml
```

{{#/tab }}
{{#/tabs }}

#### Install Infrastructure Provider Components

{{#tabs name:"tab-installation-infrastructure" tabs:"AWS,vSphere"}}
{{#tab AWS}}

Check the [AWS provider releases](https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases) for an up-to-date components file.

For more information about prerequisites, credentials management, and or permissions for AWS, visit the [getting started guide](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/getting-started.md).

```bash

> kubectl create -f https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases/download/v0.4.0/infrastructure-components.yaml

Reviewer: ```suggestion
kubectl create -f https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases/download/v0.4.1/infrastructure-components.yaml
```

I suspect we will not release v0.4.0 similar to how we handled the v0.2.0 release for capi.

```

{{#/tab }}
{{#/tabs }}

## Usage

The Cluster API resources are now installed.

![](../images/management-cluster.svg)

Now that we've got Cluster API, Bootstrap and Infrastructure resources installed,
let's proceed to create a single node cluster.

For the purpose of this tutorial, we'll name our cluster `capi-quickstart`.

{{#tabs name:"tab-usage-cluster-resource" tabs:"AWS,vSphere"}}
{{#tab AWS}}

```yaml
apiVersion: cluster.x-k8s.io/v1alpha2
kind: Cluster
metadata:
  name: capi-quickstart
spec:
  clusterNetwork:
    pods:
      cidrBlocks: ["192.168.0.0/16"]
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
    kind: AWSCluster
    name: capi-quickstart
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
kind: AWSCluster
metadata:
  name: capi-quickstart
spec:
  region: us-east-1 # Change this value to the region you want to deploy the cluster in.
  sshKeyName: default # Change this value to a valid SSH Key Pair present in your AWS Account.
```
{{#/tab }}
{{#tab vSphere}}

<aside class="note warning">

<h1>Action Required</h1>

> These examples include environment variables that you should substitute before creating the resources.

Reviewer: Should we follow a similar path for the AWS examples as well here?

Author: I don't think that's necessary, it'll be a quicker quick start for AWS :D

Reviewer: unless they are missing the pre-requisites

Author: That's fine, we call it out in some places

    apiServer:
      extraArgs:
        cloud-provider: vsphere
        cloud-config: /etc/kubernetes/vsphere.conf
      extraVolumes:
      - name: "cloud-config"
        hostPath: /etc/kubernetes/vsphere.conf
        mountPath: /etc/kubernetes/vsphere.conf
        readOnly: true
        pathType: File
    controllerManager:
      extraArgs:
        cloud-provider: vsphere
        cloud-config: /etc/kubernetes/vsphere.conf
      extraVolumes:
      - name: "cloud-config"
        hostPath: /etc/kubernetes/vsphere.conf
        mountPath: /etc/kubernetes/vsphere.conf
        readOnly: true
        pathType: File
  files:
  - path: /etc/kubernetes/vsphere.conf
    owner: root:root
    permissions: "0600"
    encoding: base64
    content: |
      ${CLOUD_CONFIG_B64ENCODED}
  users:
  - name: capv
    sudo: "ALL=(ALL) NOPASSWD:ALL"
    sshAuthorizedKeys:
    - "${SSH_AUTHORIZED_KEY}"
  preKubeadmCommands:
  - hostname "{{ ds.meta_data.hostname }}"
  - echo "::1         ipv6-localhost ipv6-loopback" >/etc/hosts
  - echo "127.0.0.1   localhost {{ ds.meta_data.hostname }}" >>/etc/hosts
  - echo "{{ ds.meta_data.hostname }}" >/etc/hostname
```
{{#/tab }}
{{#/tabs }}

> After the controlplane is up and running, we should apply the addons to install Calico.

Reviewer: ```suggestion
After the controlplane is up and running, deploy a CNI solution. Calico is used here as an example.
```

> {{#tabs name:"tab-usage-addons" tabs:"AWS,vSphere"}}

Reviewer: These tabs can be removed and we can just deploy the calico manifests directly:
```
kubectl apply -f https://docs.projectcalico.org/v3.8/manifests/calico.yaml
```

Reviewer: Just realized, to apply calico, you'll need to download and extract the kubeconfig from the secret first:
```sh
kubectl get secret capi-quickstart -o json | jq -r .data.value | base64 -d > qs.kubeconfig
kubectl --kubeconfig qs.kubeconfig apply -f https://docs.projectcalico.org/v3.8/manifests/calico.yaml
```

{{#tab AWS}}

```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-aws/v0.4.1/examples/addons.yaml
```
{{#/tab }}
{{#tab vSphere}}

```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-vsphere/v0.5.0/examples/addons.yaml
```

{{#/tab }}
{{#/tabs }}

> At this point, our control plane should be up and in `Ready` state, check the status using `kubectl get nodes`.

Reviewer: It might take a bit for the calico and coredns pods to get in a running state, might want to call out something like "After a short while, our control plane..." instead.

Finishing up, we'll create a single node _MachineDeployment_.

{{#tabs name:"tab-usage-machinedeployment" tabs:"AWS,vSphere"}}
{{#tab AWS}}

```yaml
apiVersion: cluster.x-k8s.io/v1alpha2
kind: MachineDeployment
metadata:

>   name: capi-quickstart-md-0

Reviewer: This will lead to odd MachineSet and Machine naming, maybe instead name this `capi-quickstart-worker` instead?

  labels:
    cluster.x-k8s.io/cluster-name: capi-quickstart

>     nodepool: nodepool-0

Reviewer: I'd add a comment here that this label key/value is arbitrary and just expected to be unique for a given MachineDeployment

Reviewer: We could also leave the label off here, since it isn't used by anything.

spec:
  replicas: 1
  selector:
    matchLabels:
      cluster.x-k8s.io/cluster-name: capi-quickstart

>       nodepool: nodepool-0

Reviewer: I'd add a comment here that this label key/value is arbitrary and just expected to be unique for a given MachineDeployment

  template:
    metadata:
      labels:
        cluster.x-k8s.io/cluster-name: capi-quickstart

>         nodepool: nodepool-0

Reviewer: Maybe a brief comment that the labels used here need to match the `matchLabels` above.

    spec:
      version: v1.15.3
      bootstrap:
        configRef:

>           name: capi-quickstart-md-0

Reviewer: I'd use a more generic name, otherwise it'll lead to odd names for the generated config resources.

        hostPath: /etc/kubernetes/vsphere.conf
        mountPath: /etc/kubernetes/vsphere.conf
        readOnly: true
        pathType: File
    controllerManager:
      extraArgs:
        cloud-provider: vsphere
        cloud-config: /etc/kubernetes/vsphere.conf
      extraVolumes:
      - name: "cloud-config"
        hostPath: /etc/kubernetes/vsphere.conf
        mountPath: /etc/kubernetes/vsphere.conf
        readOnly: true
        pathType: File
  files:
  - path: /etc/kubernetes/vsphere.conf
    owner: root:root
    permissions: "0600"
    encoding: base64
    content: |
      ${CLOUD_CONFIG_B64ENCODED}
  users:
  - name: capv
    sudo: "ALL=(ALL) NOPASSWD:ALL"
    sshAuthorizedKeys:
    - "${SSH_AUTHORIZED_KEY}"
  preKubeadmCommands:
  - hostname "{{ ds.meta_data.hostname }}"
  - echo "::1         ipv6-localhost ipv6-loopback" >/etc/hosts
  - echo "127.0.0.1   localhost {{ ds.meta_data.hostname }}" >>/etc/hosts
  - echo "{{ ds.meta_data.hostname }}" >/etc/hostname
```
{{#/tab }}
{{#/tabs }}

After the controlplane is up and running, deploy a CNI solution. Calico is used here as an example.

{{#tabs name:"tab-usage-addons" tabs:"Calico"}}
{{#tab Calicl}}

```bash

> kubectl apply -f https://docs.projectcalico.org/v3.8/manifests/calico.yaml

Reviewer: This is applying calico to the `kind` cluster and not the workload cluster.

Author: fixed, ptal

# Quick Start

In this tutorial we'll cover the basics of how to use Cluster API to create one or more Kubernetes clusters.

## Prerequisites

- Install and setup [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/) in your local environment.

> - A working Kubernetes cluster to be used as [management cluster](../reference/glossary.md#management-cluster).

Reviewer 2: This needs more detail for the new user + recommendation and links to KIND

## Before we start

#### Picking Infrastructure and Bootstrap Providers

The [providers section](../reference/providers.md) offers a quick reference to some providers
available and compatible with Cluster API _v1alpha2_.

<aside class="note">
<h1>Following Along vs Jumping Ahead</h1>

If you're unfamiliar with the different providers types introduced in v1alpha2
and how they interact with each other, please refer to the [user guide concepts].

[user guide concepts]: ../user/concepts.md
</aside>

> To proceed, we'll need to pick and choose a _Bootstrap_ and _Infrastructure_ provider.

Reviewer 2: On a getting started guide this needs spelling out, new folks will have no idea what this means.

Reviewer 2: Really needs a diagram IMO.

Once you've chosen which provider to use, we'll need to grab the [components]
published as part of the release.

> **NB**: Providers usually have their own getting started guide on how to generate additional resources,
> configuration, credential management, etc. Refer to their documentation before proceeding.

#### What if I don't have a management cluster?
In some cases, you might not have a Kubernetes Cluster available to be used
as management cluster for Cluster API resources.

Services like Google Cloud GKE, Amazon EKS, Azure Kuberentes Service (AKS),
or similar, are good examples of Kubernetes-as-a-service platforms

> that can be used for this purpose.

Reviewer 2: Same link to kind.  This is a quick-start so optimize the 0->working cluster.

##### Testing or development
If you're looking for a temporary cluster for testing, development, or demos,

> we suggest you to use [kind].

Reviewer 2: nah, flip this doc everything should be geared towards the newb.

```bash
GO111MODULE="on" go get sigs.k8s.io/kind@v0.5.1
kind create cluster --name=clusterapi
export KUBECONFIG="$(kind get kubeconfig-path --name="clusterapi")"
```

## Installation

Each [release](https://github.com/kubernetes-sigs/cluster-api/releases) has a list of published assets.
For the purpose of this tutorial, download the `cluster-api-components.yaml` file from the latest `v0.2.x` release.

At this point, we should have all the required [components]:
- Cluster API: `cluster-api-components.yaml`
- Bootstrap: `bootstrap-components.yaml`
- Infrastructure: `infrastructure-components.yaml`

```bash

> kubectl create -f cluster-api-components.yaml

Reviewer 2: We should provide a one shot URL to apply that just works.

#### What if I don't have a management cluster?
In some cases, you might not have a Kubernetes Cluster available to be used
as management cluster for Cluster API resources.

Services like Google Cloud GKE, Amazon EKS, Azure Kuberentes Service (AKS),
or similar, are good examples of Kubernetes-as-a-service platforms
that can be used for this purpose.

##### Testing or development
If you're looking for a temporary cluster for testing, development, or demos,
we suggest you to use [kind].

```bash
GO111MODULE="on" go get sigs.k8s.io/kind@v0.5.1
kind create cluster --name=clusterapi
export KUBECONFIG="$(kind get kubeconfig-path --name="clusterapi")"
```

## Installation

Each [release](https://github.com/kubernetes-sigs/cluster-api/releases) has a list of published assets.
For the purpose of this tutorial, download the `cluster-api-components.yaml` file from the latest `v0.2.x` release.

At this point, we should have all the required [components]:
- Cluster API: `cluster-api-components.yaml`
- Bootstrap: `bootstrap-components.yaml`
- Infrastructure: `infrastructure-components.yaml`

```bash
kubectl create -f cluster-api-components.yaml

> kubectl create -f bootstrap-components.yaml

Reviewer 3: I think at this point, It could become confusing where to get `bootstrap-components.yaml` and `infrastructure-components.yaml` from. Without explicitly calling out a provider for either is there a nice way to make reference of an out of box experience. Or would this be left up to the provider?
